### PR TITLE
Don't read past the end of input parsing regex group

### LINKED ISSUE #1052: Out of bounds reads when parsing regexp
When parsing a regexp fragment "(?", the parser will [read beyond the bounds of the input and advance](https://github.com/ruby/yarp/blob/9192d9a34db170de45ff4bbd4e86f537a4ef43e6/src/regexp.c#L382) ```parser->cursor``` beyond ```parser->end```. When memchr is eventually called, the size provided is (size_t)-1. This leads to weird Heisenbugs.

The following demonstrates:
```C
#include <yarp.h>

static unsigned char regex[] = "(?\'     ";

int main(int argc, const char **argv) {
  yp_string_list_t *capture_list = yp_string_list_alloc();
  assert(capture_list);
  yp_string_list_init(capture_list);
  // give the library 2 bytes as legitimate input: it'll read
  // past the end of the input to get a \' and trigger the bug
  yp_regexp_named_capture_group_names((const char *)regex, 2, capture_list);
  yp_string_list_free(capture_list);
  free(capture_list);
  return 0;
}
```
Build with
```
clang -ggdb3 -fsanitize=address -Iinclude $(find  src -name '*.c') regexp.c -o regexp
```

<details>
<summary>ASAN detects memchr reading through memory beyond the input buffer</summary>

```
# ./regexp 
=================================================================
==64==ERROR: AddressSanitizer: global-buffer-overflow on address 0x55e34ab6c4e9 at pc 0x55e34a9a8f71 bp 0x7ffda3a54330 sp 0x7ffda3a53b00
READ of size 64814 at 0x55e34ab6c4e9 thread T0
    #0 0x55e34a9a8f70 in memchr (/yarp/regexp+0x50f70) (BuildId: 57966dd4d0c509de3fa0f19fd37d5acfb916182d)
    #1 0x55e34aa69a8a in yp_regexp_char_find /yarp/src/regexp.c:60:38
    #2 0x55e34aa68514 in yp_regexp_parse_group /yarp/src/regexp.c:383:22
    #3 0x55e34aa67cbd in yp_regexp_parse_item /yarp/src/regexp.c:464:20
    #4 0x55e34aa67934 in yp_regexp_parse_expression /yarp/src/regexp.c:476:10
    #5 0x55e34aa67825 in yp_regexp_parse_pattern /yarp/src/regexp.c:500:13
    #6 0x55e34aa67536 in yp_regexp_named_capture_group_names /yarp/src/regexp.c:517:12
    #7 0x55e34ab02fb0 in main /yarp/regexp.c:11:3
    #8 0x7f622ca31d8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #9 0x7f622ca31e3f in __libc_start_main csu/../csu/libc-start.c:392:3
    #10 0x55e34a98e324 in _start (/yarp/regexp+0x36324) (BuildId: 57966dd4d0c509de3fa0f19fd37d5acfb916182d)

0x55e34ab6c4e9 is located 0 bytes to the right of global variable 'regex' defined in 'regexp.c:3:22' (0x55e34ab6c4e0) of size 9
  'regex' is ascii string '(?'     '
SUMMARY: AddressSanitizer: global-buffer-overflow (/yarp/regexp+0x50f70) (BuildId: 57966dd4d0c509de3fa0f19fd37d5acfb916182d) in memchr
Shadow bytes around the buggy address:
  0x0abce9565840: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
  0x0abce9565850: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
  0x0abce9565860: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
  0x0abce9565870: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
  0x0abce9565880: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
=>0x0abce9565890: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 00[01]f9 f9
  0x0abce95658a0: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
  0x0abce95658b0: f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
  0x0abce95658c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0abce95658d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0abce95658e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==64==ABORTING
```
</details>

<details>
<summary>A gdb session showing the integer overflow</summary>

```
# gdb ./regexp 
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
pwndbg: loaded 141 pwndbg commands and 43 shell commands. Type pwndbg [--shell | --all] [filter] for a list.
pwndbg: created $rebase, $ida GDB functions (can be used with print/break)
Reading symbols from ./regexp...
------- tip of the day (disable with set show-tips off) -------
Calling functions like call (void)puts("hello world") will run all other target threads for the time the function runs. If you want only the current thread to run for the function call, use set scheduler-locking on
pwndbg> b yp_regexp_char_find
Breakpoint 1 at 0x1119e2: file src/regexp.c, line 60.
pwndbg> run
Starting program: /yarp/regexp 
warning: Error disabling address space randomization: Operation not permitted
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, yp_regexp_char_find (parser=0x7ffce4133b40, value=39 '\'') at src/regexp.c:60
60	   const char *end = (const char *) memchr(parser->cursor, value, (size_t) (parser->end - parser->cursor));
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
───────────────────[ REGISTERS / show-flags off / show-compact-regs off ]────────────────────
*RAX  0x555e63b60427 (__unnamed_1+1095) ◂— 0x1d00
*RBX  0x7ffce41337a0 —▸ 0xe4133830 ◂— 0x0
*RCX  0x7ffce4133b48 —▸ 0x555e63b604e3 (regex+3) —▸ 0x2020202020 ◂— 0x0
*RDX  0x7ffce4133740 ◂— 0x1
*RDI  0x7ffce4133b40 —▸ 0x555e63b604e0 (regex) ◂— "(?'     "
*RSI  0x27
*R8   0x10001c81e6ec ◂— 0xf3f3f3f3f3f3f3f3
*R9   0x4
*R10  0x7
*R11  0x20
*R12  0x7ffce4133d58 —▸ 0x7ffce4135806 ◂— '/yarp/regexp'
*R13  0x555e63af6f40 (main) ◂— push rbp
*R14  0x555e63b45398 —▸ 0x555e63a46450 (asan.module_dtor) ◂— push rbp
*R15  0x7fcfbb313040 (_rtld_global) —▸ 0x7fcfbb3142e0 —▸ 0x555e6394c000 ◂— 0x10102464c457f
*RBP  0x7ffce4133530 —▸ 0x7ffce4133a40 —▸ 0x7ffce4133a90 —▸ 0x7ffce4133af0 —▸ 0x7ffce4133b10 ◂— ...
*RSP  0x7ffce41334c0 —▸ 0x555e63b604e1 (regex+1) ◂— 0x2020202020273f /* "?'     " */
*RIP  0x555e63a5d9e2 (yp_regexp_char_find+18) ◂— mov rax, qword ptr [rbp - 0x10]
────────────────────────────[ DISASM / x86-64 / set emulate on ]─────────────────────────────
 ► 0x555e63a5d9e2 <yp_regexp_char_find+18>    mov    rax, qword ptr [rbp - 0x10]   <__unnamed_1+1095>
   0x555e63a5d9e6 <yp_regexp_char_find+22>    add    rax, 8
   0x555e63a5d9ea <yp_regexp_char_find+26>    mov    qword ptr [rbp - 0x30], rax
   0x555e63a5d9ee <yp_regexp_char_find+30>    shr    rax, 3
   0x555e63a5d9f2 <yp_regexp_char_find+34>    cmp    byte ptr [rax + 0x7fff8000], 0
   0x555e63a5d9f9 <yp_regexp_char_find+41>    je     yp_regexp_char_find+56                <yp_regexp_char_find+56>
    ↓
   0x555e63a5da08 <yp_regexp_char_find+56>    mov    rax, qword ptr [rbp - 0x30]
   0x555e63a5da0c <yp_regexp_char_find+60>    mov    rax, qword ptr [rax]
   0x555e63a5da0f <yp_regexp_char_find+63>    mov    qword ptr [rbp - 0x48], rax
   0x555e63a5da13 <yp_regexp_char_find+67>    movsx  eax, byte ptr [rbp - 0x11]
   0x555e63a5da17 <yp_regexp_char_find+71>    mov    dword ptr [rbp - 0x3c], eax
──────────────────────────────────────[ SOURCE (CODE) ]──────────────────────────────────────
In file: /yarp/src/regexp.c
   55 }
   56 
   57 // This advances the current token to the next instance of the given character.
   58 static bool
   59 yp_regexp_char_find(yp_regexp_parser_t *parser, char value) {
 ► 60     const char *end = (const char *) memchr(parser->cursor, value, (size_t) (parser->end - parser->cursor));
   61     if (end == NULL) {
   62         return false;
   63     }
   64 
   65     parser->cursor = end + 1;
──────────────────────────────────────────[ STACK ]──────────────────────────────────────────
00:0000│ rsp 0x7ffce41334c0 —▸ 0x555e63b604e1 (regex+1) ◂— 0x2020202020273f /* "?'     " */
01:0008│     0x7ffce41334c8 —▸ 0x7ffce4133734 ◂— 0x2
02:0010│     0x7ffce41334d0 —▸ 0x7ffce4133b48 —▸ 0x555e63b604e3 (regex+3) —▸ 0x2020202020 ◂— 0x0
03:0018│     0x7ffce41334d8 —▸ 0x7ffce41336e4 ◂— 0x2
04:0020│     0x7ffce41334e0 —▸ 0x7ffce4133530 —▸ 0x7ffce4133a40 —▸ 0x7ffce4133a90 —▸ 0x7ffce4133af0 ◂— ...
05:0028│     0x7ffce41334e8 —▸ 0x7ffce41336f0 ◂— 0x2
06:0030│     0x7ffce41334f0 —▸ 0x604000000010 ◂— 0x6
07:0038│     0x7ffce41334f8 —▸ 0x7ffce4133740 ◂— 0x1
────────────────────────────────────────[ BACKTRACE ]────────────────────────────────────────
 ► 0   0x555e63a5d9e2 yp_regexp_char_find+18
   1   0x555e63a5c515 yp_regexp_parse_group+1669
   2   0x555e63a5bcbe yp_regexp_parse_item+318
   3   0x555e63a5b935 yp_regexp_parse_expression+21
   4   0x555e63a5b826 yp_regexp_parse_pattern+38
   5   0x555e63a5b537 yp_regexp_named_capture_group_names+247
   6   0x555e63af6fb1 main+113
   7   0x7fcfbafcbd90 __libc_start_call_main+128
─────────────────────────────────────────────────────────────────────────────────────────────
pwndbg> p parser->end - parser->cursor
$1 = -1
```
</details>

Tested with 9192d9a34db170de45ff4bbd4e86f537a4ef43e6.

### PR DESCRIPTION
**Pull request by stevenjohnstone**
Fixes #1052.

### CODE REVIEW

File: src/regexp.c [hunk 1]
```diff
@@ -380,7 +383,7 @@ yp_regexp_parse_group(yp_regexp_parser_t *parser) {
                 break;
             case '\'': { // named capture group
                 const char *start = ++parser->cursor;
-                if (!yp_regexp_char_find(parser, '\'')) {
+                if (yp_regexp_char_is_eof(parser) || !yp_regexp_char_find(parser, '\'')) {
```
[the remark this answers is not in the document]
Comment (stevenjohnstone) [reply src/regexp.c:386]: Yes, I think an assert would be good. Was that what you had in mind or a more general move of checks into ```yp_regexp_char_find```?

[the remark this answers is not in the document]
Comment (nobu) [reply src/regexp.c:386]: It is the latter, as I'm not sure if the behavior of `memchr` with zero length is explicitly defined

[the remark this answers is not in the document]
Comment (stevenjohnstone) [reply src/regexp.c:386]: Yes, I think you're right about ```memchr```. Something
like
```diff
diff --git a/src/regexp.c b/src/regexp.c
index 89eda531..4ea9e693 100644
--- a/src/regexp.c
+++ b/src/regexp.c
@@ -57,6 +57,9 @@ yp_regexp_char_expect(yp_regexp_parser_t *parser, char value) {
 // This advances the current token to the next instance of the given character.
 static bool
 yp_regexp_char_find(yp_regexp_parser_t *parser, char value) {
+    if (yp_regexp_char_is_eof(parser)) {
+        return false;
+    }
     const char *end = (const char *) memchr(parser->cursor, value, (size_t) (parser->end - parser->cursor));
     if (end == NULL) {
         return false;
```
?

This makes it similar to ```yp_regexp_char_accept``` and ```yp_regexp_char_expect```.

[the remark this answers is not in the document]
Comment (nobu) [reply src/regexp.c:386]: > Yes, I think you're right about `memchr`. Something like

Yes, LGTM.

> This makes it similar to `yp_regexp_char_accept` and `yp_regexp_char_expect`.

Good catch

[the remark this answers is not in the document]
Comment (stevenjohnstone) [reply src/regexp.c]: @nobu I pushed a change in line with your suggestion. Thanks!

**Review by jemmaissroff — APPROVED** [review]
Thanks for the fix!

**Pull request merged**